Modules/_threadmodule.c

```diff
@@ -1401,44 +1737,14 @@ do_start_new_thread(thread_module_state* state,
         return -1;
     }
 
-    // gh-109795: Use PyMem_RawMalloc() instead of PyMem_Malloc(),
-    // because it should be possible to call thread_bootstate_free()
-    // without holding the GIL.
-    struct bootstate *boot = PyMem_RawMalloc(sizeof(struct bootstate));
-    if (boot == NULL) {
-        PyErr_NoMemory();
+    if (ThreadHandle_start(handle, func, args, kwargs) < 0) {
         return -1;
     }
-    boot->tstate = _PyThreadState_New(interp, _PyThreadState_WHENCE_THREADING);
-    if (boot->tstate == NULL) {
-        PyMem_RawFree(boot);
-        if (!PyErr_Occurred()) {
-            PyErr_NoMemory();
-        }
-        return -1;
-    }
-    boot->func = Py_NewRef(func);
-    boot->args = Py_NewRef(args);
-    boot->kwargs = Py_XNewRef(kwargs);
-    boot->thread_is_exiting = thread_is_exiting;
-    if (thread_is_exiting != NULL) {
-        _PyEventRc_Incref(thread_is_exiting);
-    }
 
-    int err;
-    if (joinable) {
-        err = PyThread_start_joinable_thread(thread_run, (void*) boot, ident, handle);
-    } else {
-        *handle = 0;
-        *ident = PyThread_start_new_thread(thread_run, (void*) boot);
-        err = (*ident == PYTHREAD_INVALID_THREAD_ID);
-    }
-    if (err) {
-        PyErr_SetString(ThreadError, "can't start new thread");
-        PyThreadState_Clear(boot->tstate);
-        thread_bootstate_free(boot, 1);
-        return -1;
+    if (!daemon) {
+        add_to_shutdown_handles(state, handle);
```

Comment:
I think we need to add the shutdown handle earlier, before we start the thread. At this point, it's possible that the thread has already exited and tried to remove itself the shutdown handle pool. If we add the shutdown handle here, we can end up triggering the assertion `assert(self->shutdown_node.next == NULL)` in `ThreadHandle_decref`.

This can get triggered by running `test_concurrent_futures_circular_import` with `PYTHON_GIL=0`.